File: 01-programming/python/ejercicios/nivel-01-fundamentos/ejercicio-01-variables-y-tipos/main.py
# TODO: Completa el ejercicio siguiendo el enunciado de README.md.
# Sustituye cada raise NotImplementedError por la implementación correcta.


def nombre() -> str:
    # TODO: devuelve tu nombre (ej. "Ana")
    raise NotImplementedError


def ciudad() -> str:
    # TODO: devuelve tu ciudad de nacimiento (ej. "Lima")
    raise NotImplementedError


def edad() -> int:
    # TODO: devuelve tu edad (ej. 30)
    raise NotImplementedError


def estudia_programacion() -> bool:
    # TODO: devuelve True
    raise NotImplementedError


def tipo_de(valor) -> str:
    # TODO: devuelve el nombre del tipo de valor, ej. type(valor).__name__
    raise NotImplementedError


def formatear_descripcion(nombre, ciudad, edad, programacion) -> str:
    # TODO: devuelve con f-strings:
    # f"Soy {nombre}, tengo {edad} años, nací en {ciudad} y es {programacion} que estudio programación."
    raise NotImplementedError


if __name__ == "__main__":
    print(tipo_de("hola"))
    print(tipo_de(42))
    print(tipo_de(True))
    print(formatear_descripcion("Ana", "Lima", 30, True))

File: 01-programming/rust/ejercicios/nivel-01-fundamentos/ejercicio-01-variables-y-tipos/src/main.rs
fn tipo<T>(_: &T) -> &'static str {
    std::any::type_name::<T>()
}

fn main() {
    const CIUDAD: &str = "Lima";
    let nombre: &str = "Ana";
    let mut edad: u32 = 30;
    let programacion: bool = true;

    println!("nombre es de tipo {}", tipo(&nombre));
    println!("ciudad es de tipo {}", tipo(&CIUDAD));
    println!("edad es de tipo {}", tipo(&edad));
    println!("programacion es de tipo {}", tipo(&programacion));

    println!(
        "Soy {}, tengo {} años, nací en {} y es {} que estudio programación.",
        nombre, edad, CIUDAD, programacion
    );
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tipo_entero() {
        let x: i32 = 42;
        assert_eq!(tipo(&x), "i32");
    }

    #[test]
    fn test_tipo_str() {
        let s: &str = "hola";
        assert_eq!(tipo(&s), "&str");
    }

    #[test]
    fn test_tipo_u32() {
        let n: u32 = 30;
        assert_eq!(tipo(&n), "u32");
    }

    #[test]
    fn test_tipo_bool() {
        let b: bool = true;
        assert_eq!(tipo(&b), "bool");
    }
}